File: frontend/src/lib/critique.ts
// CR-056 — plan critique (ask-with-options), the client-side half.
//
// The agent NEVER auto-changes the plan. The backend attaches STRUCTURAL findings
// (duplicate / mislabel) on the proposed_action's `critique[]`; here we (a) detect
// the DATA-AWARE findings for free from the MiniReportPreview run-results the card
// already fetched (empty_dimension / single_row / identical_data — no extra
// /studio/run), (b) turn every finding into deterministic option BUTTONS, and (c)
// apply the chosen option to the in-memory draft widget list. Nothing changes until
// the user clicks; then Oluştur/Kaydet creates the ADJUSTED plan.
import type { RunResult, StudioSpec } from "@/types/studio";

export type FindingType =
  | "duplicate"
  | "mislabel"
  | "empty_dimension"
  | "single_row"
  | "identical_data";

export interface Finding {
  type: FindingType;
  widget_ids: string[];
  message: string;
}

// One editable draft widget. `key` matches the MiniReportPreview key (a widget id,
// or "report" for a single report) and the finding.widget_ids. `raw` is the original
// full widget object (dashboard/skill) preserved for create; null for a report.
export interface DraftWidget {
  key: string;
  title: string | null;
  spec: StudioSpec | null;
  raw: any | null;
}

export type OptionAction =
  | { kind: "keep_all" }
  | { kind: "remove"; key: string }
  | { kind: "keep_only"; key: string; among: string[] }
  | { kind: "retitle"; key: string };

export interface CritiqueOption {
  label: string;
  action: OptionAction;
}

const NL_UNSET = "(belirtilmemiş)";

// A stable id so a resolved finding stays hidden and its badges clear.
export function findingId(f: Finding): string {
  return `${f.type}:${[...f.widget_ids].sort().join(",")}`;
}

// Strip a "%" / "(%)" label from a mislabeled title (retitle resolution). Drops a
// trailing "(%)"/"( % )" group, a stray "%", and the word "yüzde", then tidies
// leftover empty parens and double spaces.
export function stripPercentLabel(title: string | null | undefined): string {
  let t = String(title ?? "");
  t = t.replace(/\(\s*%\s*\)/g, " "); // "(%)" / "( % )"
  t = t.replace(/%/g, " ");
  t = t.replace(/\byüzde\b/gi, " ");
  t = t.replace(/\(\s*\)/g, " "); // empty parens left behind
  t = t.replace(/\s+/g, " ").trim();
  return t;
}

// Mirror of the backend structural signature: sorted metrics + ordered dims + a
// stable filter key. Used to tell identical_data (DIFFERENT spec, same data) from a
// plain duplicate (same signature — already flagged by the backend).
export function signatureOf(spec: StudioSpec | null | undefined): string {
  const s: any = spec ?? {};
  const metrics = [...((s.metrics as string[]) ?? [])].map(String).sort();
  const dims = [...((s.dimensions as string[]) ?? [])].map(String);
  const filters = ((s.filters as any[]) ?? [])
    .filter((f) => f && typeof f === "object")
    .map((f) => [String(f.field), String(f.op), String(f.value)] as const)
    .sort((a, b) => (a.join("|") < b.join("|") ? -1 : 1));
  return JSON.stringify([metrics, dims, filters]);
}

// A canonical fingerprint of a run RESULT (rows + totals), order-independent, so two
// widgets that render the same numbers are caught even with different specs.
function resultFingerprint(r: RunResult): string {
  const rows = (r.rows ?? [])
    .map((row) => JSON.stringify([row.dims ?? {}, row.metrics ?? {}]))
    .sort();
  return JSON.stringify([rows, r.totals?.metrics ?? {}]);
}

// The distinct values a widget's FIRST breakdown dimension takes across the result
// rows (null/blank folded to one "(belirtilmemiş)" bucket).
function distinctDimValues(spec: StudioSpec, r: RunResult): Set<string> {
  const dim = (spec.dimensions ?? [])[0];
  const out = new Set<string>();
  if (!dim) return out;
  for (const row of r.rows ?? []) {
    const v = row.dims?.[dim];
    out.add(v == null || v === "" ? NL_UNSET : String(v));
  }
  return out;
}

// Detect the data-aware findings from the preview widgets + their fetched results.
// Only widgets with a result are considered (results arrive as previews mount).
export function detectDataFindings(
  widgets: { key: string; title: string | null; spec: StudioSpec }[],
  resultByKey: Record<string, RunResult | undefined>
): Finding[] {
  const findings: Finding[] = [];

  // empty_dimension / single_row (per widget).
  for (const w of widgets) {
    const r = resultByKey[w.key];
    if (!r) continue;
    const dims = w.spec.dimensions ?? [];
    const nRows = (r.rows ?? []).length;
    if (dims.length >= 1) {
      const distinct = distinctDimValues(w.spec, r);
      const onlyUnset = distinct.size === 1 && distinct.has(NL_UNSET);
      if (distinct.size <= 1) {
        findings.push({
          type: "empty_dimension",
          widget_ids: [w.key],
          message: `‘${w.title ?? "Bu widget"}’ kırılımında veri yok${
            onlyUnset ? " (tümü “belirtilmemiş”)" : " (tek değer)"
          }; bu kırılım bir şey katmıyor.`,
        });
      }
    } else if (w.spec.viz !== "kpi" && nRows <= 1) {
      // A breakdown-less table/chart that came back with ≤1 row.
      findings.push({
        type: "single_row",
        widget_ids: [w.key],
        message: `‘${w.title ?? "Bu widget"}’ tek satır döndürdü; tablo/grafik bir şey katmıyor.`,
      });
    }
  }

  // identical_data — DIFFERENT spec, IDENTICAL result. Group by result fingerprint
  // but only pair widgets whose structural signature DIFFERS (same-signature twins
  // are the backend's `duplicate`, not a data-level twin).
  const byPrint = new Map<string, { key: string; title: string | null; sig: string }[]>();
  for (const w of widgets) {
    const r = resultByKey[w.key];
    if (!r || (r.rows ?? []).length === 0) continue; // no empty-vs-empty "twins"
    const print = resultFingerprint(r);
    const arr = byPrint.get(print) ?? [];
    arr.push({ key: w.key, title: w.title, sig: signatureOf(w.spec) });
    byPrint.set(print, arr);
  }
  for (const group of byPrint.values()) {
    const distinctSigs = new Set(group.map((g) => g.sig));
    if (group.length > 1 && distinctSigs.size > 1) {
      const titles = group.map((g) => `‘${g.title ?? "Adsız"}’`).join(" ile ");
      findings.push({
        type: "identical_data",
        widget_ids: group.map((g) => g.key),
        message: `${titles} farklı tanımlı ama aynı veriyi üretiyor.`,
      });
    }
  }

  return findings;
}

// The deterministic option buttons for a finding, labeled from the current draft
// titles. Resolution (trim/retitle) happens client-side via applyOption.
export function optionsFor(f: Finding, titleByKey: Record<string, string | null>): CritiqueOption[] {
  const titleOf = (k: string) => titleByKey[k] ?? "Adsız";
  switch (f.type) {
    case "duplicate":
    case "identical_data":
      return [
        { label: "İkisini de tut", action: { kind: "keep_all" } },
        ...f.widget_ids.map((k) => ({
          label: `Sadece “${titleOf(k)}”`,
          action: { kind: "keep_only" as const, key: k, among: f.widget_ids },
        })),
      ];
    case "empty_dimension":
    case "single_row":
      return [
        { label: "Tut", action: { kind: "keep_all" } },
        { label: "Kaldır", action: { kind: "remove", key: f.widget_ids[0] } },
      ];
    case "mislabel":
      return [
        { label: "Başlığı düzelt", action: { kind: "retitle", key: f.widget_ids[0] } },
        { label: "Olduğu gibi bırak", action: { kind: "keep_all" } },
      ];
  }
}

// Badge shown inline on an affected preview widget.
export function badgeForType(t: FindingType): string {
  switch (t) {
    case "duplicate":
    case "identical_data":
      return "Yinelenen";
    case "empty_dimension":
    case "single_row":
      return "Veri yok";
    case "mislabel":
      return "Etiket uyumsuz";
  }
}

// Apply a chosen option to the draft widget list — PURE (returns a new list). This
// is the ONLY thing that changes the plan, and only on the user's click.
export function applyOption(draft: DraftWidget[], action: OptionAction): DraftWidget[] {
  switch (action.kind) {
    case "keep_all":
      return draft;
    case "remove":
      return draft.filter((d) => d.key !== action.key);
    case "keep_only":
      return draft.filter((d) => d.key === action.key || !action.among.includes(d.key));
    case "retitle":
      return draft.map((d) =>
        d.key === action.key
          ? {
              ...d,
              title: stripPercentLabel(d.title),
              raw: d.raw ? { ...d.raw, title: stripPercentLabel(d.raw.title ?? d.title) } : d.raw,
            }
          : d
      );
  }
}

File: backend/app/services/studio/critique.py
"""CR-056 — plan critique (structural, at compile). ADVISORY ONLY: this module
DETECTS issues in a proposed report/pano/beceri plan; it NEVER mutates the plan.

The founder's chosen behavior is *always ask with options* — so the agent surfaces
what it noticed and the user's click (client-side, deterministic) is what trims or
retitles the draft. Nothing here changes a widget; ``build_critique`` reads the
widgets and returns findings the ``ProposedActionCard`` renders as a summary + inline
badges + option buttons.

A finding = ``{type, widget_ids[], message (tr-TR)}``. The option buttons and the
trimming/retitle are owned by the FE (one place, so both structural and data-aware
findings resolve identically). Two structural checks live here because they need no
data (authoritative at compile, cheaply testable):

  * ``duplicate`` — two data widgets with the SAME signature
    ``(sorted metric ids, dims, filter key)`` → identical tables/charts (the DGN
    "Kalem Kalem Gider = Maliyet Kategorilerinin Dağılımı" case). Mirrors the CR-052
    ``_filter_key`` idea so widgets that share metric+dim but differ in filters are
    NOT twins.
  * ``mislabel`` — the title disagrees with the metric type: a "%"-titled widget over
    a ``currency`` metric (the "Maliyet Dağılımı (%)" plotting ₺), or a ₺/$-titled
    widget over a ``percent`` metric.

The data-aware checks (``empty_dimension`` / ``single_row`` / ``identical_data``) run
in the FE from the ``MiniReportPreview`` run-results already fetched — no extra
``/studio/run`` calls.
"""
from app.services.studio.catalog import METRICS


def _filter_key(spec: dict) -> tuple:
    """A stable, hashable key for a widget's filters (CR-052 idea), so two widgets
    with the same metric+dim but DIFFERENT filters are distinct, not twins."""
    out = []
    for f in (spec or {}).get("filters") or []:
        if isinstance(f, dict):
            out.append((str(f.get("field")), str(f.get("op")), str(f.get("value"))))
    return tuple(sorted(out))


def _signature(spec: dict) -> tuple:
    """The structural identity of a data widget: sorted metric ids + ordered dims +
    filter key. Two widgets with the same signature render the same table/chart."""
    spec = spec or {}
    metrics = tuple(sorted(str(m) for m in (spec.get("metrics") or [])))
    dims = tuple(str(d) for d in (spec.get("dimensions") or []))
    return (metrics, dims, _filter_key(spec))


def _metric_types(spec: dict) -> set:
    """The catalog ``type`` of each of the widget's metrics (unknown ids dropped)."""
    return {METRICS[m]["type"] for m in (spec or {}).get("metrics") or [] if m in METRICS}


def _title(w: dict) -> str:
    return str(w.get("title") or "").strip()


def _is_data_widget(w: dict) -> bool:
    spec = w.get("spec")
    return isinstance(spec, dict) and bool(spec.get("metrics"))


def build_critique(widgets: list[dict]) -> list[dict]:
    """Detect structural findings in a plan's widgets. ``widgets`` is a list of
    ``{id, title, spec}`` (text/report widgets — no metric spec — are ignored).
    Returns a possibly-empty findings list. NEVER mutates ``widgets`` (advisory)."""
    data = [w for w in (widgets or []) if _is_data_widget(w)]
    findings: list[dict] = []

    # duplicate — group by structural signature; any group of >1 is a twin set.
    by_sig: dict[tuple, list[dict]] = {}
    for w in data:
        by_sig.setdefault(_signature(w["spec"]), []).append(w)
    for group in by_sig.values():
        if len(group) > 1:
            titles = " ile ".join(f"‘{_title(w) or 'Adsız'}’" for w in group)
            findings.append({
                "type": "duplicate",
                "widget_ids": [str(w["id"]) for w in group],
                "message": f"{titles} aynı veriyi gösteriyor (aynı metrik ve kırılım).",
            })

    # mislabel — the title implies "%" but the metric is ₺ (or the reverse).
    for w in data:
        title = _title(w)
        types = _metric_types(w["spec"])
        if not types:
            continue
        if "%" in title and "percent" not in types:
            findings.append({
                "type": "mislabel",
                "widget_ids": [str(w["id"])],
                "message": (
                    f"‘{title}’ başlığı yüzde (%) ima ediyor ama metrik para birimi (₺). "
                    "Başlığı düzeltmek ister misin?"
                ),
            })
        elif ("₺" in title or "$" in title) and types == {"percent"}:
            findings.append({
                "type": "mislabel",
                "widget_ids": [str(w["id"])],
                "message": (
                    f"‘{title}’ başlığı para birimi ima ediyor ama metrik yüzde (%). "
                    "Başlığı düzeltmek ister misin?"
                ),
            })

    return findings


def critique_summary(findings: list[dict]) -> str:
    """A one-line tr-TR summary for the tool result (so the agent relays what it
    noticed and asks). ``''`` when there are no findings (the message is unchanged)."""
    if not findings:
        return ""
    return (
        "Taslağı hazırladım. Şunları fark ettim: "
        + " ".join(f["message"] for f in findings)
        + " Ne yapmamı istersin?"
    )
